src/lib/array.jl

```diff
@@ -279,7 +279,10 @@ end
   enumerate(xs), back
 end
 
-@adjoint Iterators.Filter(f, x) = pullback(filter, f, collect(x))
+function _pullback(cx::AContext, ::Type{<:Iterators.Filter}, f, x)
```

Comment:
Note on why these are `_pullback` instead of `@adjoint`: `@adjoint` tries to do some clever postprocessing of the pullback return which results in mismatched results (differential arg tuple off by one). Using `_pullback` is slightly more code and requires a bit more care, but it removes a few layers from the call stack and brings us closer to removing `@adjoint` completely.

Reply:
Yes, when Miha and I were trying to get rid of Zygote's own types we likewise removed some uses of `@adjoint` and wrote the `_pullback` overloads directly